### PR TITLE
[Feature](top-n) support topn filter on vscan node

### PR DESCRIPTION
**Pull request by BiteTheDDDDt**
## Proposed changes
support topn filter on vscan node

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

File: be/src/vec/utils/util.hpp [hunk 1]
```diff
@@ -169,6 +170,63 @@ inline std::string remove_suffix(const std::string& name, const std::string& suf
...
+    if (is_column_const(*column)) {
+        change_null_to_true(assert_cast<const ColumnConst*>(column.get())->get_data_column_ptr());
+    } else if (column->is_nullable()) {
+        auto* nullable =
+                const_cast<ColumnNullable*>(assert_cast<const ColumnNullable*>(column.get()));
+        auto* __restrict data = assert_cast<ColumnUInt8*>(nullable->get_nested_column_ptr().get())
+                                        ->get_data()
+                                        .data();
+        auto* __restrict null_map = const_cast<uint8_t*>(nullable->get_null_map_data().data());
+        for (size_t i = 0; i < column->size(); ++i) {
```
Comment (HappenLee) [inline be/src/vec/utils/util.hpp:193]: call `column->size()` in for loop cause not SIMD,  not call the fun in for loop

File: be/src/vec/exprs/vtopn_pred.h [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    static Status create_vtopn_pred(SlotDescriptor* slot_desc, int source_node_id,
+                                    vectorized::VExprSPtr& expr) {
+        TExprNode node;
+        node.__set_node_type(TExprNodeType::FUNCTION_CALL);
+        node.__set_type(create_type_desc(PrimitiveType::TYPE_BOOLEAN));
+        node.__set_is_nullable(slot_desc->is_nullable());
+        expr = vectorized::VTopNPred::create_shared(node, source_node_id);
+
+        expr->add_child(VSlotRef::create_shared(slot_desc));
```
Comment (HappenLee) [inline be/src/vec/exprs/vtopn_pred.h:52]: think about the logic? maybe here not need be a slot desc? can execute a func?

File: be/src/vec/exprs/vtopn_pred.h [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+        std::vector<size_t> arguments = {(size_t)slot_id, (size_t)topn_value_id};
+
+        size_t num_columns_without_result = block->columns();
+        block->insert({nullptr, _data_type, _expr_name});
+        RETURN_IF_ERROR(_function->execute(nullptr, *block, arguments, num_columns_without_result,
+                                           block->rows(), false));
+        *result_column_id = num_columns_without_result;
+
+        if (is_nullable()) {
```
Comment (HappenLee) [inline be/src/vec/exprs/vtopn_pred.h:100]: maybe only null first need always not filter ?

**Comment by BiteTheDDDDt** [comment]
run buildall

File: be/src/pipeline/exec/scan_operator.cpp [hunk 4]
```diff
@@ -1300,6 +1301,27 @@ Status ScanLocalState<Derived>::_init_profile() {
     return Status::OK();
 }
 
+template <typename Derived>
+Status ScanLocalState<Derived>::_get_topn_filters(RuntimeState* state) {
+    for (auto id : get_topn_filter_source_node_ids()) {
+        const auto& pred = state->get_query_ctx()->get_runtime_predicate(id);
+        if (!pred.inited()) {
```
Comment (HappenLee) [inline be/src/pipeline/exec/scan_operator.cpp:1308]: DCHECK

**Review by englefly — APPROVED** [review]

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**